### PR TITLE
test: setup E2E tests for ORAS

### PR DESCRIPTION
**Pull request by qweeah**
This PR introduces E2E testing based on [ginkgo](https://github.com/onsi/ginkgo) and [gomega](https://github.com/onsi/gomega). E2E tests will be run against ORAS binary built during `Build CLI` step and latest [oras-distribution container](https://github.com/oras-project/distribution/pkgs/container/registry) with below changes:

- Build CI action: Added start & stop steps for backend distribution
- Test: added E2E tests running together with unit tests during `make test`
  - Two test suites: `TestORASCommands` for single command tests and `TestORASScenarios` for scenario-based tests
  - Testing utilities: added command execution helper types and funcs, as well as a output matcher package.
  
Now the PR also contains sample e2e tests for:
1) `login` & `logout` commands, as well as failure checks for all commands after logging out .
2) helper doc check for `manifest fetch` command

To run the tests, a guide can be found in [here](https://hackmd.io/AvWPi6FnTOyPTwHJwpwodw).
related to #523

### CODE REVIEW

File: test/e2e/utils/init.go [hunk 1]
```diff
@@ -0,0 +1,78 @@
...
+	if err := (registry.Reference{Registry: Host}).ValidateRegistry(); Host == "" || err != nil {
+		panic(err)
+	}
+	var _ = BeforeSuite(func() {
+		OrasPath = os.Getenv("ORAS_PATH")
+
+		if filepath.IsAbs(OrasPath) {
+			// test against OrasPath directly
+			fmt.Printf("Testing based on pre-built binary locates in %q\n", OrasPath)
+		} else if workspacePath := os.Getenv("GITHUB_WORKSPACE"); filepath.IsAbs(OrasPath) && workspacePath != "" {
```
Comment (chloeyin) [inline test/e2e/utils/init.go:53]: This is empty branch
```suggestion
		} else if workspacePath := os.Getenv("GITHUB_WORKSPACE"); !filepath.IsAbs(OrasPath) && workspacePath != "" {
```

File: .github/workflows/build.yml [hunk 2]
```diff
@@ -24,8 +24,21 @@ jobs:
...
+          docker stop oras-e2e || true
+          docker run -d -p 5000:5000 --rm --name oras-e2e \
+            --mount type=bind,source=$GITHUB_WORKSPACE/test/e2e/testdata/config-example-with-extensions.yml,target=/etc/docker/registry/config.yml \
+            --mount type=bind,source=$GITHUB_WORKSPACE/test/e2e/testdata/passwd_bcrypt,target=/etc/docker/registry/passwd \
+            ghcr.io/oras-project/registry:latest
+      - name: Run Tests
+        run: |
+          trap docker stop oras-e2e || true ERR
+          make test
+          docker stop oras-e2e || true
```
Comment (shizhMSFT) [inline .github/workflows/build.yml:38]: What's the difference between `docker stop` and `docker kill`?

File: test/e2e/utils/file.go [hunk 3]
```diff
@@ -0,0 +1,66 @@
...
+
+func ReadFullFile(path string) ([]byte, error) {
+	fp, err := os.OpenFile(path, os.O_RDONLY, 0666)
+	if err != nil {
+		return nil, err
+	}
+
+	defer fp.Close()
+	return io.ReadAll(fp)
+}
```
Comment (shizhMSFT) [inline test/e2e/utils/file.go:30]: This method looks exactly the same as [io.ReadFile](https://pkg.go.dev/os#ReadFile).

File: .github/workflows/build.yml (hunk 2 above)
Comment (qweeah) [reply .github/workflows/build.yml:51]: looks like `docker stop` will wait 10s and then kill? Should use kill to save time.

File: test/e2e/utils/file.go (hunk 3 above)
Comment (qweeah) [reply test/e2e/utils/file.go]: It's **os**.ReadFile. Thanks for pointing out!!!

**Comment by qweeah** [comment]
Closing this, will add another PR for testing utilities

**Pull request closed without merging**